src/matrixToDataFrame.c

```diff
@@ -0,0 +1,312 @@
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include "tibble.h"
+
+static SEXP pairlist_shallow_copy(SEXP p)
+{
+    SEXP attr, q, s;
+    int nprot = 0;
+
+    PROTECT(attr = Rf_cons(CAR(p), R_NilValue)); nprot++;
+    q = attr;
+    SET_TAG(q, TAG(p));
+    p = CDR(p);
+
+    while (!Rf_isNull(p)) {
+        PROTECT(s = Rf_cons(CAR(p), R_NilValue)); nprot++;
+        SETCDR(q, s);
+        q = CDR(q);
+        SET_TAG(q, TAG(p));
+        p = CDR(p);
+    }
+
+    UNPROTECT(nprot);
+    return attr ;
+}
+
+static void copy_attributes(SEXP out, SEXP data)
+{
+    SEXP att = ATTRIB(data);
+    if (!Rf_isNull(att)){
+        SET_ATTRIB(out, pairlist_shallow_copy(ATTRIB(data))) ;
+    }
+    SET_OBJECT(out, OBJECT(data));
+    if (IS_S4_OBJECT(data)) {
+        SET_S4_OBJECT(out);
+    }
+}
+
+// same as copy_attributes but without names
+static void copy_most_attributes(SEXP out, SEXP data)
+{
+    copy_attributes(out, data);
+    Rf_setAttrib(out, R_NamesSymbol, R_NilValue);
+}
+
+static void get_dim(SEXP x, R_xlen_t *nrowptr, R_xlen_t *ncolptr)
+{
+    SEXP dim;
+    R_xlen_t nrow = 0, ncol = 0;
+    int nprot = 0;
+
+    PROTECT(dim = getAttrib(x, R_DimSymbol)); nprot++;
+    if (dim == R_NilValue || XLENGTH(dim) != 2) {
+        error("`x` is not a matrix");
+    }
+
+    switch (TYPEOF(dim)) {
+    case INTSXP:
+        nrow = (R_xlen_t)INTEGER(dim)[0];
+        ncol = (R_xlen_t)INTEGER(dim)[1];
+        break;
+
+    case REALSXP:
+        nrow = (R_xlen_t)REAL(dim)[0];
+        ncol = (R_xlen_t)REAL(dim)[1];
+        break;
+
+    default:
+        error("`x` is not a matrix");
+        break;
+    }
+
+    if (nrowptr) {
+        *nrowptr = nrow;
+    }
+    if (ncolptr) {
+        *ncolptr = ncol;
+    }
+    UNPROTECT(nprot);
+}
+
+static SEXP get_names(SEXP x, R_xlen_t ncol)
+{
+    SEXP dimnames, colnames;
+    R_xlen_t i;
+    char *buf;
+    int nprot = 0, ndigit;
+
+    // check for column names, use them if present
+    PROTECT(dimnames = getAttrib(x, R_DimNamesSymbol)); nprot++;
+    if (TYPEOF(dimnames) == VECSXP && XLENGTH(dimnames) == 2) {
+        colnames = VECTOR_ELT(dimnames, 1);
+        if (TYPEOF(colnames) == STRSXP) {
+            goto Return;
+        }
+    }
+
+    // otherwise, allocate new names
+    PROTECT(colnames = allocVector(STRSXP, ncol)); nprot++;
+
+    // if we really wanted to save a few bytes:
+    //
+    //  ndigit = (ncol == 0) ? 1 : 1 + floor(log10((double)ncol));
+    //
+    // but there's no need to get fancy since log10(2^64) = 19.266
+    ndigit = 20;
+    buf = R_alloc(1 + ndigit + 1, 1); // V + (number) + NUL
+
+    for (i = 0; i < ncol; i++) {
+        sprintf(buf, "V%"PRIu64, (uint64_t)(i + 1));
+        SET_STRING_ELT(colnames, i, mkCharCE(buf, CE_UTF8));
+    }
+
+Return:
+    UNPROTECT(nprot);
+    return colnames;
+}
+
+static SEXP get_rownames(SEXP x, R_xlen_t nrow)
+{
+    SEXP dimnames, rownames;
+    int nprot = 0;
+
+#ifndef TIBBLE_DROP_ROWNAMES
```

Comment:
I'm fine with keeping existing row names here. Could you please add a test?

Reply:
Actually, no: let's keep row names out: https://github.com/tidyverse/tibble/issues/288#issuecomment-328855126. Happy to support a `rownames` argument in `as_tibble.matrix()`, which then could be extended to `as_tibble.data.frame()`. (The concept of keys I brought up in the linked reference is orthogonal, we need to store the data in a column anyway, with or without keys.)

Reply:
I left them in, but feel free to edit. It seems cleaner to me to handle all the row names logic in just one function. Just do

```
as_tibble.matrix <- function(x, ...) {
    as_tibble(matrixToDataFrame(x), ...)
}
```